game-core/src/main/java/games/strategy/triplea/attachments/TechAbilityAttachment.java

```diff
@@ -102,25 +110,75 @@ public TechAbilityAttachment(final String name, final Attachable attachable, fin
     super(name, attachable, gameData);
   }
 
-  // setters and getters
+  private UnitType getUnitType(final String value) throws GameParseException {
+    return Optional.ofNullable(getData().getUnitTypeList().getUnitType(value))
+        .orElseThrow(() -> new GameParseException("No unit called:" + value + thisErrorMsg()));
+  }
+
+  @VisibleForTesting
+  String[] splitAndValidate(final String name, final String value) throws GameParseException {
+    final String[] stringArray = value.split(":");
+    if (value.isEmpty() || stringArray.length > 2) {
+      throw new GameParseException(
+          String.format("%s cannot be empty or have more than two fields %s", name, thisErrorMsg()));
+    }
+    return stringArray;
+  }
+
+  private void applyCheckedValue(
+      final String name,
+      final String value,
+      final BiConsumer<UnitType, Integer> putter) throws GameParseException {
+    final String[] s = splitAndValidate(name, value);
+    putter.accept(getUnitType(s[1]), getInt(s[0]));
+  }
+
+  private static int sumIntegerMap(final Function<TechAbilityAttachment, IntegerMap<UnitType>> mapper,
+      final UnitType ut,
+      final PlayerID player,
+      final GameData data) {
+    return TechTracker.getCurrentTechAdvances(player, data).stream()
+        .map(TechAbilityAttachment::get)
+        .filter(Objects::nonNull)
+        .map(mapper)
+        .mapToInt(m -> m.getInt(ut))
+        .sum();
+  }
+
+  private static int sumNumbers(
```

Comment:
@RoiEXLab this looks to be the main functionality, it would be the thing to test IMO. Would you have objection to adding a test for these specifically? I think you could do so pretty reasonable by elevating scope with `@VisibleForTesting` and some mock inputs. 

That would get us some pretty solid coverage.

I would go one step further personally and move these functions to a strategy object which is then injected to TechAttachmentAbility, that would then allow us to control validation and we could then check if we get an exception or not when validation fails. That would reduce dependency on exact exception types and error messages, and would simplify testing of tech attachment ability to just two cases